Fix Bug 1471514 - Add top-of-funnel metrics for fxa

ericawright opened the pull request:
@sarracini @vladikoff

~To Test: change line 32 from "https://accounts.firefox.com/metrics-flow" to "https://latest.dev.lcip.org/metrics-flow"
Currently there is a CORs issue with accounts.firefox.com and the changes required will hopefully be live in a few days.~

re: https://github.com/mozilla/fxa-content-server/issues/6267
re: https://github.com/mozilla/fxa-content-server/issues/5882

piatra commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -12,12 +12,26 @@ export class _StartupOverlay extends React.PureComponent {
+      }
+    };
+    req.onerror = e => {
+      console.error("fxa error", e);
```
Should we report these to telemetry?

ericawright replied:
sure!

ericawright replied:
@vladikoff, would fxa use/want this data?

vladikoff replied:
@ericawright not for now, worst case we can look at telemetry I guess!

ericawright replied:
@vladikoff that's what I mean, if I store this event in telemety would you use it, since AS doesn't need it.

vladikoff replied:
@ericawright oh, if it is easy enough and you have other errors already sent to telemetry let's do it! We can later look if the request fails for users

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -12,12 +12,26 @@ export class _StartupOverlay extends React.PureComponent {
   }
 
   initScene() {
+    let req = new XMLHttpRequest();
```
Is there any reason you can't use the regular fetch API here instead of XMLHttpRequest? https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -34,6 +48,11 @@ export class _StartupOverlay extends React.PureComponent {
     }, 400);
   }
 
+  setFormFields(r) {
```
Why a new function? Why not just set the state in the callback of the request when you have the response handy?

vladikoff commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -12,12 +12,26 @@ export class _StartupOverlay extends React.PureComponent {
+      console.error("fxa error", e);
+    };
+
+    req.open("GET", "https://accounts.firefox.com/metrics-flow");
```
can we use any existing FxA settings or setup a setting in activity stream that can be changed in `about:config` to alter this url?

ericawright replied:
I found identity.fxaccounts.remote.root, but it sounds like you'd want to be able to change the whole thing.
I'll make a new one, how about identity.fxaccounts.remote.metrics?
(I also noticed there are urls in about:config for legal/terms and legal/privacy - I'll add those to a different PR)

ericawright replied:
though perhaps adding a new pref for this url should also go in a separate PR since this one will need to be uplifted

ericawright replied:
I take that back, I'll use a pref and fallback to the hardcoded url if there is no pref, then if the pref code doesn't get uplifted this will at least still work, and then enterprise folks or others who wish to block this could.
sound okay?

ericawright commented:
@sarracini I fixed the failing test, not sure if I've done it correctly with a  fetch stub. Also, added a pref for the metrics endpoint. I would prefer the pref gets set in product and not from AS - though that may be an irrational concern.

vladikoff commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -13,11 +15,20 @@ export class _StartupOverlay extends React.PureComponent {
+        this.setState({flowId: response.flowId, flowBeginTime: response.flowBeginTime});
+      })
+      .catch(error => {
+        this.props.dispatch(ac.UndesiredEvent({type: "FXA_METRICS_ERROR", error}));
```
awesome! 👍

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -2,6 +2,8 @@ import {FormattedMessage, injectIntl} from "react-intl";
 import {connect} from "react-redux";
 import React from "react";
+/* eslint-disable import/no-commonjs */
+// const fetch = require("node-fetch");
```
leftover?

vladikoff commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -80,6 +89,8 @@ export class _StartupOverlay extends React.PureComponent {
                 <input name="entrypoint" type="hidden" value="activity-stream-firstrun" />
```
@ericawright this form above has `accounts.firefox.com` hardcoded vs using `this.props.fxa_metrics_endpoint` like the fetch script. Is it possible to fix that?

ericawright replied:
yeah I think that's already in the prefs, I'll find it

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -13,11 +13,20 @@ export class _StartupOverlay extends React.PureComponent {
   }
 
   initScene() {
+    window.fetch(this.props.fxa_endpoint)
```
This is actually not working as intended. If you log `this.props.fxa_endpoint` at this point, it's undefined. Therefore this actually throws all the time. You can put a log in the `.catch` and you can see that it throws:
`TypeError: "undefined is not a valid URL."`
The pref you're looking for is not ready in the constructor of this component unless you put it in the prerendered data's initial prefs values. The prefs get sent too late for this, and so you never get any of the data you're looking for, and it attempts to do a fetch on `undefined` which will throw

ericawright replied:
I'm not getting that error, I was getting a cors error (since I accidentally lost the /metrics-flow portion of the URl when changing the contents of the pref.

sarracini replied:
![screen shot 2018-07-04 at 11 07 13 am](https://user-images.githubusercontent.com/7219526/42284428-796be60a-7f7a-11e8-8ca3-613127d94ec8.png)

ericawright commented on lib/ActivityStream.jsm:
```diff
@@ -170,6 +170,10 @@ const PREFS_CONFIG = new Map([
   ["asrouter.snippetsUrl", {
     title: "A custom URL for the AS router snippets",
     value: ""
+  }],
```
seems to need to be here and in the prerender, else it gets erased

sarracini replied:
You can remove this with the changes mentioned above

sarracini commented on common/PrerenderData.jsm:
```diff
@@ -63,7 +63,8 @@ this.PrerenderData = new _PrerenderData({
     "sectionOrder": "topsites,topstories,highlights",
-    "collapsed": false
+    "collapsed": false,
+    "fxaccounts.endpoint": "https://accounts.firefox.com"
```
So this won't be the right approach either. Remove this from here, remove the pref from ActivityStream.jsm, and add a look up for the pref: identity.fxaccounts.remote.root with a fallback value in PrefsFeed after values.platform:

```js
values.fxa_endpoint = Services.prefs.getStringPref("identity.fxaccounts.remote.root", "https://accounts.firefox.com");
```

Then in StartupOverlay.jsx connect Prefs:
```js
const getState = state => ({fxa_endpoint: state.Prefs.values.fxa_endpoint});
export const StartupOverlay = connect(getState)(injectIntl(_StartupOverlay));
```

And then you can use this.props.fxa_endpoint directly in the component

vladikoff replied:
Amazing!

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -15,12 +15,24 @@ export class _StartupOverlay extends React.PureComponent {
   }
 
-  initScene() {
+  async initScene() {
```
Call this in `componentDidMount`, I just noticed this was in the constructor and it shouldn't be there

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -15,12 +15,24 @@ export class _StartupOverlay extends React.PureComponent {
-  initScene() {
+  async initScene() {
+    try {
+      let response = await fetch(`${this.props.fxa_endpoint}/metrics-flow`);
```
Do this fetch stuff in componentWillUpdate, that's when the props from the connected Prefs will be ready. Any earlier than that and our store isn't ready.
Also add a flag to make sure we're not doing multiple fetchs when we update:

```js
  async componentWillUpdate() {
    try {
      if (!this.props.fxa_endpoint || this.didFetch) {
        return;
      }
      this.didFetch = true;
      const response = await fetch(`${this.props.fxa_endpoint}/metrics-flow`);
      if (response.status === 200) {
        const {flowId, flowBeginTime} = await response.json();
        this.setState({flowId, flowBeginTime});
      }
    } catch (error) {
      this.props.dispatch(ac.OnlyToMain({type: at.TELEMETRY_UNDESIRED_EVENT, data: {value: "FXA_METRICS_ERROR"}}));
    }
  }
```

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -28,6 +41,18 @@ export class _StartupOverlay extends React.PureComponent {
     }, 10);
   }
 
+  async setFxaMetrics() {
```
Again, no need for a different function here, just adds an extra bind in the constructor that you don't need.

sarracini commented on content-src/components/Base/Base.jsx:
```diff
@@ -152,7 +152,7 @@ export class BaseContent extends React.PureComponent {
           </main>
         </div>
-        {this.props.isFirstrun && <StartupOverlay />}
+        {this.props.isFirstrun && <StartupOverlay fxa_endpoint={prefs["fxaccounts.endpoint"]} />}
```
you don't need this prop here, you're getting the prop from the connected state in StartupOverlay.jsx

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -12,11 +12,24 @@ export class _StartupOverlay extends React.PureComponent {
+  }
+
+  componentWillUpdate() {
+    if (this.props.fxa_endpoint && !this.didFetch) {
```
You need to define this.didFetch in the constructor with value of false

ericawright replied:
sure, I figured undefined worked well enough

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -12,11 +12,24 @@ export class _StartupOverlay extends React.PureComponent {
+  componentWillUpdate() {
+    if (this.props.fxa_endpoint && !this.didFetch) {
+      this.didFetch = true;
+      this.setFxaMetrics();
```
do you need to `await` on this function here since it's async?

sarracini replied:
if you remove this function and just put the logic in componentWillUpdate then you can disregard this comment

ericawright replied:
I was going to put the logic in the componentWillUpdate, but then it runs multiple times being async, the flag does not get set in time to stop the second time from running. 
I was also considering, on failure, doing a setTimout and running this again up to 2 or 3 times (circumventing the flag) just in case there is a temporary error, like a bad connection or something.

sarracini commented on lib/PrefsFeed.jsm:
```diff
@@ -102,6 +102,7 @@ this.PrefsFeed = class PrefsFeed {
     // computed in main process
     values.isPrivateBrowsingEnabled = PrivateBrowsingUtils.enabled;
     values.platform = AppConstants.platform;
+    values.fxa_endpoint = Services.prefs.getStringPref("identity.fxaccounts.remote.root", "https://accounts.firefox.com");
```
a comment here would be helpful about why this is necessary

sarracini commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -15,8 +15,29 @@ export class _StartupOverlay extends React.PureComponent {
+    if (this.props.fxa_endpoint && !this.didFetch) {
+      try {
+        this.didFetch = true;
+        let response = await fetch(`/metrics-flow`);
```
supposed to be 
```js
const response = await fetch(`${this.props.fxa_endpoint}/metrics-flow`);`
```

Mardak commented on content-src/components/StartupOverlay/StartupOverlay.jsx:
```diff
@@ -15,8 +15,29 @@ export class _StartupOverlay extends React.PureComponent {
+    if (this.props.fxa_endpoint && !this.didFetch) {
+      try {
+        this.didFetch = true;
+        const response = await fetch(`${this.props.fxa_endpoint}/metrics-flow`);
```
Looks like this is causing test failure timeout:
TEST-UNEXPECTED-FAIL | browser/base/content/test/urlbar/browser_urlbar_blanking.js | Found a tab after previous test timed out: about:welcome -
https://treeherder.mozilla.org/logviewer.html#?job_id=187870714&repo=try&lineNumber=1633

Where the test loads gInitialPages:
https://searchfox.org/mozilla-central/source/browser/base/content/browser.js#321

The test times out with about:welcome open and status bar showing "Performing TLS handshake to 127.0.0.1"

Mardak replied:
Ah, during testing, it's set to {server} so 127.0.0.1, and the request silently fails:

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://127.0.0.1:8888//metrics-flow. (Reason: CORS request did not succeed).

Mardak replied:
The following works because it looks like the test/metrics-flow endpoint just never resolves in time for the test:
```js
const controller = new AbortController();
setTimeout(() => controller.abort(), 5000);
const response = await fetch(`${this.props.fxa_endpoint}/metrics-flow`, {signal: controller.signal});
```

Unclear how long we should wait before aborting (and triggering a FXA_METRICS_ERROR…

ericawright replied:
@mardak would you prefer something like that, or to edit the test?

sarracini replied:
It's not just a single test, so you won't be able to edit the test

ericawright replied:
true, it would fail anywhere this page is used wouldn't it

Mardak commented:
Reverted in https://github.com/mozilla/activity-stream/commit/245cf97a5847d7885328347f6335da32c5549008 for the mochitest failures
